flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala

```diff
@@ -0,0 +1,237 @@
+//    val defaultValue = Some(1.0)
+//  }
+//
+//  def setMiniBatchFraction(fraction: Double): GradientDescent = {
+//    parameterMap.add(MiniBatchFraction, fraction)
+//    this
+//  }
+
+  /** Performs one iteration of Stochastic Gradient Descent using mini batches
+    *
+    * @param data A Dataset of LabeledVector (label, features) pairs
+    * @param currentWeights A Dataset with the current weights to be optimized as its only element
+    * @return A Dataset containing the weights after one stochastic gradient descent step
+    */
+  private def SGDStep(data: DataSet[(LabeledVector)], currentWeights: DataSet[WeightVector]):
+    DataSet[WeightVector] = {
+
+    // TODO: Sample from input to realize proper SGD
+    data.map {
+      new GradientCalculation
+    }.withBroadcastSet(currentWeights, WEIGHTVECTOR_BROADCAST).reduce {
+      (left, right) =>
+        val (leftGradientVector, leftCount) = left
+        val (rightGradientVector, rightCount) = right
+
+        BLAS.axpy(1.0, leftGradientVector.weights, rightGradientVector.weights)
+        (new WeightVector(
+          rightGradientVector.weights,
+          leftGradientVector.intercept + rightGradientVector.intercept),
+          leftCount + rightCount)
+    }.map {
+      new WeightsUpdate
+    }.withBroadcastSet(currentWeights, WEIGHTVECTOR_BROADCAST)
+  }
+
+  /** Provides a solution for the given optimization problem
+    *
+    * @param data A Dataset of LabeledVector (label, features) pairs
+    * @param initWeights The initial weights that will be optimized
+    * @return The weights, optimized for the provided data.
+    */
+  override def optimize(data: DataSet[LabeledVector], initWeights: Option[DataSet[WeightVector]]):
+  DataSet[WeightVector] = {
+    // TODO: Faster way to do this?
+    val dimensionsDS = data.map(_.vector.size).reduce((a, b) => b)
+
+    val numberOfIterations: Int = parameterMap(Iterations)
+
+
+    val initialWeightsDS: DataSet[WeightVector] = initWeights match {
+      //TODO(tvas): Need to figure out if and where we want to pass initial weights
+      case Some(x) => x
+      case None => createInitialWeightVector(dimensionsDS)
+    }
+
+    // We need the weights vector to initialize the regularization value.
+    val initWeightsVector = initialWeightsDS.collect()(0).weights
+
+    val initRegVal = parameterMap(RegularizationType)
+      .applyRegularization(initWeightsVector, 0.0, parameterMap(RegularizationParameter))._2
+
+    // TODO: Is there a way to initialize the regularization parameter without collect()?
+    // The following code should give us a dataset with the initial reg. parameter, but we still
+    // need to call collect to retrieve it. Question is: call collect here, or on the weights as we
+    // currently do?
+    //val initRegValue: DataSet[Double] = initialWeights.map {x => parameterMap(RegularizationType)
+    //  .applyRegularization(x.weights, 0.0, parameterMap(RegularizationParameter))._2}
+    //  .reduce(_ + _)
```

Comment:
I think you don't need a reduce operation here, because the `initialWeights` data set only contains one element.